pkg/chart/repo/repo.go

```diff
@@ -0,0 +1,284 @@
+	indexURL := parsed.String()
+
+	data, err := doFetch(indexURL)
+	if err != nil {
+		return fmt.Errorf("failed to fetch %q: %v", indexURL, err)
+	}
+
+	if _, err := loadIndex(data); err != nil {
+		return fmt.Errorf("failed to load index file: %v", err)
+	}
+
+	if err := r.cache.Store("index.yaml", data); err != nil {
+		return fmt.Errorf("failed to cache index.yaml: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Repo) ResolveVersion(chart, version string) (*repo.ChartVersion, error) {
+	versions, err := r.ChartVersions(chart, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get chart versions: %s", err)
+	}
+
+	if len(versions) == 0 {
+		return nil, repo.ErrNoChartVersion
+	}
+
+	var highestver *repo.ChartVersion
+	var lasterr error
+	for _, ver := range versions {
+		if _, lasterr = r.Fetch(ver); lasterr == nil {
+			highestver = ver
+			break
+		}
+		if _, lasterr := r.LoadCached(ver); lasterr == nil {
+			highestver = ver
+			break
+		}
+	}
+	if highestver == nil {
+		return nil, fmt.Errorf("failed to resolve version for chart %s %s: last error: %s", chart, version, lasterr)
+	}
+
+	return highestver, nil
+}
+
+func (r *Repo) ChartVersions(name, version string) (repo.ChartVersions, error) {
+	data, err := r.cache.Fetch("index.yaml")
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch index.yaml from cache: %v", err)
+	}
+	index, err := loadIndex(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load index file: %v", err)
+	}
+
+	vs, ok := index.Entries[name]
+	if !ok {
+		return nil, repo.ErrNoChartName
+	}
+	if len(vs) == 0 {
+		return nil, repo.ErrNoChartVersion
+	}
+
+	var constraint *semver.Constraints
+	if len(version) == 0 {
+		constraint, _ = semver.NewConstraint("*")
+	} else {
+		var err error
+		constraint, err = semver.NewConstraint(version)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	versions := make([]*repo.ChartVersion, 0, len(vs))
+	for _, ver := range vs {
+		test, err := semver.NewVersion(ver.Version)
+		if err != nil {
+			continue
+		}
+		if !constraint.Check(test) {
+			continue
+		}
+		versions = append(versions, ver)
+	}
+
+	return versions, nil
+}
+
+func (r *Repo) LoadCached(cv *repo.ChartVersion) (*chart.Chart, error) {
+	filename := chart2file(cv.Name, cv.Version)
+	data, err := r.cache.Fetch(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	return loadChart(data)
+}
+
+func (r *Repo) Fetch(cv *repo.ChartVersion) (*chart.Chart, error) {
+	if cv == nil {
+		return nil, fmt.Errorf("chart version is nil, can not proceed")
+	}
+	if len(cv.URLs) == 0 {
+		return nil, fmt.Errorf("chart %q has no downloadable URLs", cv.Name)
+	}
+
+	// copy-paste from Helm's chart_downloader.go
+	chartURL, err := url.Parse(cv.URLs[0])
+	if err != nil {
+		return nil, fmt.Errorf("invalid chart URL format: %v", cv.URLs[0])
+	}
+
+	// If the URL is relative (no scheme), prepend the chart repo's base URL
+	if !chartURL.IsAbs() {
+		repoURL, _ := url.Parse(r.url) // URL is validated before
```

Comment:
I think doing this URL parsing when creating the repo gives a stronger guarantee that it is in fact valid. Otherwise, it's relying on the detail that it's done in `RefreshIndex`, which may or may not hold true forever.